lib/public/Http/Client/IPromise.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @copyright Copyright (c) 2023, Joas Schilling <coding@schilljs.com>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCP\Http\Client;
+
+/**
+ * A wrapper around Guzzle's PromiseInterface
+ * @see \GuzzleHttp\Promise\PromiseInterface
+ * @since 28.0.0
+ */
+interface IPromise {
+	/**
+	 * @since 28.0.0
+	 */
+	public const PENDING = 'pending';
+	/**
+	 * @since 28.0.0
+	 */
+	public const FULFILLED = 'fulfilled';
+	/**
+	 * @since 28.0.0
+	 */
+	public const REJECTED = 'rejected';
+
+	/**
+	 * Appends fulfillment and rejection handlers to the promise, and returns
+	 * a new promise resolving to the return value of the called handler.
+	 *
+	 * @param callable $onFulfilled Invoked when the promise fulfills. Gets an \OCP\Http\Client\IResponse passed in as argument
+	 * @param callable $onRejected  Invoked when the promise is rejected. Gets an \Exception passed in as argument
+	 *
+	 * @return IPromise
+	 * @since 28.0.0
+	 */
+	public function then(
+		callable $onFulfilled = null,
+		callable $onRejected = null,
+	): IPromise;
+
+	/**
+	 * Get the state of the promise ("pending", "rejected", or "fulfilled").
+	 *
+	 * The three states can be checked against the constants defined on
+	 * PromiseInterface: PENDING, FULFILLED, and REJECTED.
+	 *
+	 * @return string
```

Comment:
```suggestion
	 * @return self::STATE_*
```

if you rename the consts